R/bundlePackage.R

```diff
@@ -246,37 +242,3 @@ performPackratSnapshot <- function(bundleDir, verbose = FALSE) {
 snapshotLockFile <- function(appDir) {
   file.path(appDir, "packrat", "packrat.lock")
 }
-
-# Return TRUE when the source indicates that a package was installed from
-# source or comes from a source control system. This indicates that we will
-# not have a repostory URL; location is recorded elsewhere.
-isSCMSource <- function(source) {
-  tolower(source) %in% c("github", "gitlab", "bitbucket", "source")
-}
-
-# generate a random name prefixed with "repo_".
-random.repo.name <- function() {
-  paste("repo_", paste(sample(LETTERS, 8, replace = TRUE), collapse = ""), sep = "")
-}
-
-# Given a list of optionally named repository URLs, return a list of
-# repository URLs where each element is named. Incoming names are preserved.
-# Un-named repositories are given random names.
-name.all.repos <- function(repos) {
```

Comment:
A couple of things about repository naming:

1. The previous implementation set "Source" to the name of the repository in the "repos" list. This implementation uses the "Source" from Packrat, which is often the `Repository` field in the `DESCRIPTION`.

2. Because we were using the `repos` names, this `name.all.repos` function helped make sure that all repository URLs had names, for cases where folks use bare URLs:

    ```r
    options(repos = "https://cran.rstudio.com/")
    ```

There was some uncertainty and variability about how the `Repository` field was populated for different package repositories (CRAN, Package Manager, others), and we saw some package repositories not include `Repository`, and manifests end up with empty `Source`. I don't recall the details, but maybe using the `repos` names meant that `CustomCRANLikeRepository` wasn't in the manifest?

Connect wants a nonempty "Source" value.
If Connect sees one "Source" name with multiple URLs, that is converted into:

```r
options(repos=c(NAME="firstURL", NAME_2="secondURL", ...))
```

Related, prior work: https://github.com/rstudio/rsconnect/pull/309

Reply:
How does `Source` end up empty? I don't see any [cases in packrat](https://github.com/rstudio/packrat/blob/v0.9.0/R/pkg.R#L328) where it would not be set.

For the case of [custom repositories](https://github.com/rstudio/packrat/blob/v0.9.0/R/pkg.R#L385-L391), isn't the the repository url going to be an ok `Source` as it's already unique? 

Or are you saying that if `findRepo()` is called with an unknown `pkg` (i.e. it's not a package in a known repository) then we should return something other than `NA`?

Reply:
Custom repositories: this change uses the `Source` coming out of Packrat, which is either the `Repository` field from the `DESCRIPTION` or `CustomCRANLikeRepository`, right? That looks to be fine.

Reply:
Connect errs if it receives an empty `Source` field in the manifest; that's a constraint on whatever we do here.

Reply:
You're right -- we should not see empty source, since `checkBundlePackages` will have already erred if any incoming `Source` is NA.

By ignoring the names in `options("repos")`, there may be a subtle problem.

Let's say that you have some custom repository and configure it with `options(repos = c(CORPORATE = "URL"))`. This is convention within your company.

That `CORPORATE` repository name may not be in the `DESCRIPTION` files (because you are using some custom repository). Coming out of Packrat, the "Source" today will have a value of `CORPORATE`, but with this change, will become `CustomCRANLikeRepository` (I think this is accurate from reading the code).

Your Connect configuration may contain an `RPackageRepository` configuration which rewrites `CORPORATE` repository URLs. https://docs.posit.co/connect/admin/appendix/configuration/#RPackageRepository

> This isn't a perfect strategy (because the repository names are client-controlled), but is how the feature functions.
> We want to improve this capability; it's fairly restrictive today.

Maybe: If the `DESCRIPTION` does not name the repository and we see a value of `CustomCRANLikeRepository`, it should be rewritten to the name given to the `options("repos")` repository URL.